### ISSUE TITLE
Custom jvp of dominant eigensolver yields wrong vjp behavior

### ISSUE THREAD

**Issue by buwantaiji** [google/jax#7302]
I'm using jax to test the AD of dominant eigensolver. As an example, the function `symeig` below returns only the lowest eigenvalue and corresponding eigenvector of a real symmetric matrix.

```python
from jax.config import config
config.update("jax_enable_x64", True)
import jax
from jax import grad
import jax.numpy as jnp

def symeig(A):
    eigvals, eigvectors = jax.scipy.linalg.eigh(A)
    eigval, eigvector = eigvals[0], eigvectors[:, 0]
    return eigval, eigvector
```
This function is itself differentiable, since it just wraps the full diagonalization routine `jax.scipy.linalg.eigh`. Nevertheless, I manually implemented the jvp rule, which is well known as the standard perturbation theory in the context of quantum mechanics.

```python
# jvp version 1: "normal" tensor operations using the whole spectrum.
symeig_jvp1 = jax.custom_jvp(symeig)
@symeig_jvp1.defjvp
def symeig_jvp1_jvp(primals, tangents):
    A, = primals
    dA, = tangents
    eigvals, eigvectors = jax.scipy.linalg.eigh(A)
    eigval, eigvector = primal_out = eigvals[0], eigvectors[:, 0]

    deigval = eigvector.dot(dA).dot(eigvector)
    F = eigvals - eigvals[:, None] + jnp.eye(A.shape[0])
    F = 1. / F - jnp.eye(A.shape[0])
    deigvector = eigvectors.dot( (F * eigvectors.T.dot(dA).dot(eigvectors))[:, 0] )
    tangent_out = deigval, deigvector

    return primal_out, tangent_out
```

On the other hand, there exists an alternative implementation of the jvp rules, in which the tangent of the eigenvector is obtained by solving a low-rank linear system. Such an approach turns out to be much more efficient when the matrix to be diagonalized is huge. The reason is that it requires no access to full eigen-spectrum of the matrix, and one thus can resort to sparse eigensolvers like the Lanczos algorithm. See, e.g., [https://journals.aps.org/prb/abstract/10.1103/PhysRevB.101.245139](https://journals.aps.org/prb/abstract/10.1103/PhysRevB.101.245139) for details.

```python
# jvp version 2: linear system solver.
symeig_jvp2 = jax.custom_jvp(symeig)
@symeig_jvp2.defjvp
def symeig_jvp2_jvp(primals, tangents):
    A, = primals
    dA, = tangents
    eigval, eigvector = primal_out = symeig_jvp2(A)

    deigval = eigvector.dot(dA).dot(eigvector)
    b = -dA.dot(eigvector)
    b -= eigvector * eigvector.dot(b)
    Aprime = A - eigval * jnp.eye(A.shape[0])
    deigvector, _ = jax.scipy.sparse.linalg.cg(Aprime, b, tol=1e-12)
    tangent_out = deigval, deigvector

    return primal_out, tangent_out
```

The equivalence of the two jvp implementations can be confirmed by the test function `test_jvp` below. However, when I try to perform a reverse-mode AD computation by using jax's transpose mechanism from jvp to vjp rules, the second jvp implementation above yields incorrect gradient, which is shown in the test function `test_grad` below.

```python
def test_jvp(A, dA):
    (eigval, eigvector), (deigval, deigvector) = jax.jvp(symeig, (A,), (dA,))
    (eigval1, eigvector1), (deigval1, deigvector1) = jax.jvp(symeig_jvp1, (A,), (dA,))
    (eigval2, eigvector2), (deigval2, deigvector2) = jax.jvp(symeig_jvp2, (A,), (dA,))

    assert jnp.allclose(eigval, eigval1) and jnp.allclose(eigval, eigval2)
    assert jnp.allclose(eigvector, eigvector1) and jnp.allclose(eigvector, eigvector2)
    assert jnp.allclose(deigval, deigval1) and jnp.allclose(deigval, deigval2)
    assert jnp.allclose(deigvector, deigvector1) and jnp.allclose(deigvector, deigvector2)

def test_grad(A, v):
    def make_loss(symeig, v):
        def loss_fn(A):
            _, eigvector = symeig(A)
            loss = jnp.abs(eigvector**2 - v**2).sum()
            return loss
        return loss_fn

    grad0 = grad(make_loss(symeig, v))(A)
    grad1 = grad(make_loss(symeig_jvp1, v))(A)
    grad2 = grad(make_loss(symeig_jvp2, v))(A)
    grad0 = 0.5 * (grad0 + grad0.T)
    grad1 = 0.5 * (grad1 + grad1.T)
    grad2 = 0.5 * (grad2 + grad2.T)
    assert jnp.allclose(grad0, grad1)
    assert jnp.allclose(grad0, grad2)    # AssertionError here!

key = jax.random.PRNGKey(0)
N = 100
A = jax.random.normal(key, (N, N))
A = 0.5 * (A + A.T)
key, subkey = jax.random.split(key)
dA = jax.random.normal(subkey, (N, N))
dA = 0.5 * (dA + dA.T)

test_jvp(A, dA)    # passed

key, subkey = jax.random.split(key)
v = jax.random.normal(subkey, (N,))
v /= jnp.linalg.norm(v)

test_grad(A, v)    # did not pass
```

I do not know much about the details of jax's transpose mechanism from jvp to vjp, but it seems to fail on the jvp version 2 above. Any clarification or correction is appreciated, since this implementation can be very useful for certain kinds of applications..

**Comment by froystig** [comment]
cc @j-towns and @shoyer, who have thought about `eigh` autodiff before. Also cc @mblondel, who has mentioned (in conversation) an issue before about transpose-of-jvp giving unexpected results.

**Comment by shoyer** [comment]
This seems like something that definitely _should_ work -- these partial eigenvalue solvers are quite useful for large scale problems.

I tried decreasing N and switching to `np.testing.assert_allclose` (to show the size of errors). It's clear that there's a large discrepancy here, though.

It might be worth looking at JAX's own test suite for eigh grads to see if that is more informative, e.g., something based of:
https://github.com/google/jax/blob/29187a3317c649c72bfeb06b6816dc09eb476050/tests/linalg_test.py#L393

There are definitely a lot of subtleties with defining and testing eigenvector derivatives, e.g., degeneracies, functions defined only on symmetric inputs, ambiguous phase, etc. Overall I'm still not sure if this is the best approach, or if perhaps it is better to define custom derivatives for better defined functions like matrix-valued functions / power-series.

**Comment by shoyer** [comment]
In an attempt to narrow this down, I wrote a third version using `jnp.linalg.solve` instead of `cg`:
```python
# jvp version 3: direct linear solve.
symeig_jvp3 = jax.custom_jvp(symeig)
@symeig_jvp3.defjvp
def symeig_jvp3_jvp(primals, tangents):
    A, = primals
    dA, = tangents
    eigval, eigvector = primal_out = symeig_jvp3(A)

    deigval = eigvector.dot(dA).dot(eigvector)
    b = -dA.dot(eigvector)
    b -= eigvector * eigvector.dot(b)
    Aprime = A - eigval * jnp.eye(A.shape[0])
    deigvector = jnp.linalg.solve(Aprime, b)
    tangent_out = deigval, deigvector

    return primal_out, tangent_out
```

Interesting, this function now fails the JVP test, too!

Looking at the calculation more carefully, I don't think your JVP rule is correct. The problem is that Aprime is a not full rank, so the linear solve is not well defined. It's been a while since I looked into this carefully, but as I recall you can either do a pseudo-inverse solve or solve some enlarged linear system to make this well defined.

In summary -- please double check your math, and then let's see if there's still a JAX bug  :)

**Comment by buwantaiji** [comment]
@shoyer Thanks for the response! As you have noticed, the linear system involved in jvp version 2 is low-rank. The point, however, is that the linear system can be unambiguously specified.

For the eigenvalue problem A v_i = lambda_i v_i of a real symmetric matrix A, the jvp rule can be derived from perturbation theory and reads (assuming no degeneracy):
![jvp perturbation](https://user-images.githubusercontent.com/25216189/126493168-3923c0f3-87f2-42d1-8301-18f807ea325b.png)
These formulas are implemented in jvp version 1. The essential point is that the tangent dv_0 for the desired eigenvector can also be obtained by solving the following low-rank linear system:
![jvp linear system](https://user-images.githubusercontent.com/25216189/126493630-b0fa44d9-cade-44fa-a103-f8a8dd39762b.png)
The second equation constrains dv_0 to be in the subspace spanned by all eigenvectors of A **other than v_0**, which has dimension N-1. As a result, this problem is "effectively full-rank", and can be solved by using iterative linear system solvers like `jax.scipy.sparse.linalg.cg`, as implemented in jvp version 2 above. The explicit solvers like `jnp.linalg.solve`, on the other hand, are not designed to tackle this kind of problem. This is why the version 3 code you wrote did not pass the jvp test.

**Comment by wangleiphy** [comment]
Yes. @buwantaiji's math is correct. Also see https://math.mit.edu/~stevenj/18.336/adjoint.pdf: texts above eq. (6) and footnote 7.

**Comment by shoyer** [comment]
OK, well regardless for debugging it would be helpful to make a version of this custom JVP rule that does not use `cg`. The would help narrow whether the issue is with `custom_jvp` or the custom autodiff machinery used inside `cg` itself (via `lax.custom_linear_solve`). It may also be JAX's rule for transposed linear solves implicitly assumes full rank, even if you don't need that to correctly calculate JVPs.

**Comment by buwantaiji** [comment]
I made the following version of jvp rule using matrix pseudo-inverse `jnp.linalg.pinv`. It passes both tests `test_jvp` and `test_grad`.
```python
# jvp version 3: pseudo-inverse.
symeig_jvp3 = jax.custom_jvp(symeig)
@symeig_jvp3.defjvp
def symeig_jvp3_jvp(primals, tangents):
    A, = primals
    dA, = tangents
    eigval, eigvector = primal_out = symeig_jvp3(A)

    deigval = eigvector.dot(dA).dot(eigvector)
    b = -dA.dot(eigvector)
    b -= eigvector * eigvector.dot(b)
    Aprime = A - eigval * jnp.eye(A.shape[0])
    Aprime_pinv = jnp.linalg.pinv(Aprime, rcond=1e-15)
    deigvector = Aprime_pinv.dot(b)
    tangent_out = deigval, deigvector

    return primal_out, tangent_out
```
The issue seems to be attributed to transpose of iterative solvers like `cg`.

**Comment by shoyer** [comment]
This is very helpful, thank you!
